core/src/overlay.rs

```diff
@@ -0,0 +1,239 @@
+use std::{cell::RefCell, rc::Rc};
+
+use ribir_geom::{Size, ZERO_SIZE};
+use ribir_macros::{fn_widget, Query};
+
+use crate::prelude::*;
+use crate::{
+  context::{AppCtx, BuildCtx, LayoutCtx},
+  state::{StateWriter, Stateful, Writer},
+  widget::{BoxClamp, Render, WidgetBuilder, WidgetId, WidgetTree},
+};
+
+/// Use to control the overlay, including position and whether to show or not.
+/// Return when the Overlay insert into the window.
+pub struct OverlayControl {
+  is_show: bool,
+  anchor: Anchor,
+}
+
+impl OverlayControl {
+  /// show the overlay.
+  pub fn show(&mut self) { self.is_show = true; }
+
+  /// remove the overlay from the widget tree.
+  pub fn remove(&mut self) { self.is_show = false; }
+
+  /// check if the overlay is showing.
+  pub fn is_show(&self) -> bool { self.is_show }
+
+  /// set the global anchor to the overlay.
+  pub fn set_anchor(&mut self, global_anchor: Anchor) { self.anchor = global_anchor; }
+
+  /// get the global anchor of the overlay.
+  pub fn get_anchor(&self) -> Anchor { self.anchor }
+}
+
+pub trait OverlayBuilder: Sized {
+  type Value: WidgetBuilder + 'static;
+  fn build(&mut self, ctrl: Writer<OverlayControl>) -> Self::Value;
+
+  fn map<O>(mut self, f: impl Fn(Writer<OverlayControl>, Self::Value) -> O) -> impl OverlayBuilder
+  where
+    O: WidgetBuilder + 'static,
+  {
+    OverlayCreator {
+      builder: move |ctrl: Writer<OverlayControl>| f(ctrl.clone_writer(), self.build(ctrl)),
+    }
+  }
+
+  fn with_mask_layer(self, brush: impl Into<Brush>, tap_remove: bool) -> impl OverlayBuilder
```

Comment:
We should provide a style to make the overlay more flexible, not just a mask layer. I prefer to remove the `with_mask_layer` and support the style when we refactor the widgets library.